clippy_lints/src/utils/higher.rs

```diff
@@ -114,13 +114,36 @@ pub fn range(expr: &hir::Expr) -> Option<Range> {
 
 /// Checks if a `let` decl is from a `for` loop desugaring.
 pub fn is_from_for_desugar(decl: &hir::Decl) -> bool {
+    // This will detect plain for-loops without an actual variable binding:
+    //
+    // ```
+    // for x in some_vec {
+    //   // do stuff
+    // }
+    // ```
     if_let_chain! {[
         let hir::DeclLocal(ref loc) = decl.node,
         let Some(ref expr) = loc.init,
         let hir::ExprMatch(_, _, hir::MatchSource::ForLoopDesugar) = expr.node,
     ], {
         return true;
     }}
+
+    // This detects a variable binding in for loop to avoid `let_unit_value`
```

Comment:
Cannot see it matching this for-loop. `is_from_for_desugar` returns false for the following:

```
#![feature(plugin)]
#![plugin(clippy)]

fn main() {
    let b = vec![Foo { a: 42 }];

    for Foo { a } in b {
        println!("a: {}", a);
    }
}

struct Foo {
    a: u32
}
```

<details><summary>The `{:#?}` output of decl in above example</summary>

```
Spanned {
    node: DeclLocal(
        Local {
            pat: pat(8: b),
            ty: None,
            init: Some(
                expr(26: <[_]>::into_vec(box [Foo{a: 42,}]))
            ),
            id: NodeId(
                7
            ),
            hir_id: HirId {
                owner: DefIndex(
                    3
                ),
                local_id: ItemLocalId(
                    1
                )
            },
            span: ./example.rs:5:5: 5:32,
            attrs: ThinVec(
                None
            ),
            source: Normal
        }
    ),
    span: ./example.rs:5:5: 5:33
}
```

</details>